cranelift/isle/isle/src/serialize.rs

```diff
@@ -636,82 +661,80 @@ impl<'a> Decomposition<'a> {
-
-        // All valid candidates sort greater than this so the kind of control flow
-        // doesn't matter.
-        let mut best = Candidate::with_invalid_key(HasControlFlow::Match(0.try_into().unwrap()));
-
-        // We sorted the candidates in reverse order of their sort key, so the
-        // first one is usually the best option we have. However, equality-
-        // constraint candidates are over-approximated. Those identify how many
-        // rules have some kind of equality constraint involving a particular
-        // binding site, but we actually need to find the rules with an
-        // equality constraint on a particular _pair_ of binding sites. Those
-        // are the intersection of two candidates' rules, so the upper bound on
-        // the number of matching rules is whichever candidate is smaller.
-        while let Some((candidate, rest)) = unmatched.split_first() {
-            unmatched = rest;
-            if candidate.key() <= best.key() {
+        let (normals, equals) = candidates.split_at_mut(normals);
+
+        // Find the best normal candidate.
+        let mut best = None;
+        for candidate in normals.iter() {
+            if best.as_ref() < Some(candidate) {
+                best = Some(candidate.clone());
+            }
+        }
+
+        // Everything before this took O(n) time.
+
+        // Equality-constraint candidates are over-approximated. They identify
+        // how many rules have some kind of equality constraint involving a
+        // particular binding site, but we actually need to find the rules
+        // with an equality constraint on a particular _pair_ of binding
+        // sites. Those are the intersection of two candidates' rules, so the
+        // upper bound on the number of matching rules is whichever candidate
+        // is smaller. Do an O(n log n) sort before the O(n^2) all-pairs loop
+        // so we can do branch-and-bound style pruning.
+        equals.sort_unstable_by(|x, y| y.cmp(x));
+        let mut equals = &equals[..];
+        while let Some((x, rest)) = equals.split_first() {
+            equals = rest;
+            if Some(x) < best.as_ref() {
                 break;
             }
-            match candidate.kind {
-                HasControlFlow::Match(_) | HasControlFlow::Loop(_) => return Some(candidate.kind),
-                HasControlFlow::Equal(a, _) => {
-                    for other in unmatched.iter().chain(matched.iter()) {
-                        if let HasControlFlow::Equal(b, _) = other.kind {
-                            let mut new = Candidate {
-                                // `partition` will find the intersection of
-                                // these two candidates' rules, and the best
-                                // case is one is a subset of the other.
-                                count: candidate.count.min(other.count),
+            if let HasControlFlow::Equal(x_id, _) = x.kind {
```

Comment:
This version of the logic is really close to complete, I think; the main thing that's still a bit unsatisfying to me is the dual use of `HasControlFlow::Equal(x, y)` as a sort of sentinel for "check anything related to `x` (and normally inserted as `Equal(x, x)` but the second tuple field is ignored here) and as an actual pair to emit a check for below. Could we instead have two variants on `HasControlFlow` for the two cases, something like `EqualityInvolving(x)` and `Equal(x, y)`? Then we can insert the former as a candidate elsewhere, and we can panic if we see it when converting to `ControlFlow`. (Alternately, two different enums, one for the domain of each stage of processing; but I'm fine with keeping the union of the domains in one enum for expediency here if you want.)